src/uproot/behaviors/TBranch.py

```diff
@@ -689,6 +689,200 @@ def foreach(start):
+        raise NotImplementedError()
+
+    real_options = dict(options)
+    if "num_workers" not in real_options:
+        real_options["num_workers"] = 1
+    if "num_fallback_workers" not in real_options:
+        real_options["num_fallback_workers"] = 1
+
+    filter_branch = uproot._util.regularize_filter(filter_branch)
+
+    all_arrays = []
+    file_all_keys = []
+    count = 0
+
+    for file_path, object_path in files:
+        hasbranches = _regularize_object_path(
+            file_path, object_path, custom_classes, allow_missing, real_options
+        )
+        if hasbranches is not None:
+            count += 1
+            new_keys = hasbranches.keys(
+                recursive=recursive,
+                filter_name=filter_name,
+                filter_typename=filter_typename,
+                filter_branch=filter_branch,
+                full_paths=full_paths,
+            )
+
+            all_arrays.append(hasbranches)
+
+            if expressions is not None:
+                file_all_keys.append(expressions)
+            else:
+                file_all_keys.append(new_keys)
+
+    to_keys = list(set.intersection(*map(set, file_all_keys)))
+    common_keys = sorted(to_keys, key=lambda k: file_all_keys[0].index(k))
+
+    if count == 0:
+        raise ValueError(
+            "allow_missing=True and no TTrees found in\n\n    {}".format(
+                "\n    ".join(
+                    "{"
+                    + "{}: {}".format(
+                        repr(f.file_path if isinstance(f, HasBranches) else f),
+                        repr(f.object_path if isinstance(f, HasBranches) else o),
+                    )
+                    + "}"
+                    for f, o in files
+                )
+            )
+        )
+
+    if len(common_keys) == 0:
+        raise ValueError(
+            "TTrees in\n\n    {}\n\nhave no TBranches in common".format(
+                "\n    ".join(
+                    "{"
+                    + "{}: {}".format(
+                        repr(f.file_path if isinstance(f, HasBranches) else f),
+                        repr(f.object_path if isinstance(f, HasBranches) else o),
+                    )
+                    + "}"
+                    for f, o in files
+                )
+            )
+        )
+
+    arg_keys = [expressions]
+    if expressions is not None:
+        common_keys = (
+            arg_keys[0]
+            if len(arg_keys[0]) == 1
+            else arg_keys
+            if len(str(arg_keys[0])) == len(arg_keys[0])
+            else expressions
+        )
+    else:
+        common_keys = common_keys
+
+    out_dict = {}
+    for key in common_keys:
+        out_dask = []
+
+        for branch in all_arrays:
+            shape = (branch[key].num_entries,)
+            d_dtype = numpy.array(branch[key]).dtype
+            d_array = dask.delayed(branch[key].array)(library="np")
+            dasked_d = da.from_delayed(d_array, shape=shape, dtype=d_dtype)
```

Comment:
I have modified the function, such that obtaining the `dtype` is done using information that's available before reading actually occurs.